external/storm-redis/src/main/java/org/apache/storm/redis/trident/state/RedisMapState.java

```diff
@@ -0,0 +1,327 @@
+        this.serializer = serializer;
+        this.keyFactory = keyFactory;
+    }
+
+    public List<T> multiGet(List<List<Object>> keys) {
+        if (keys.size() == 0) {
+            return Collections.emptyList();
+        }
+        if (Strings.isNullOrEmpty(this.options.hkey)) {
+            String[] stringKeys = buildKeys(keys);
+            Jedis jedis = null;
+            try {
+                jedis = jedisPool.getResource();
+                List<String> values = jedis.mget(stringKeys);
+                return deserializeValues(keys, values);
+            } finally {
+                if (jedis != null) {
+                    jedisPool.returnResource(jedis);
+                }
+            }
+        } else {
+            Jedis jedis = null;
+            try {
+                jedis = jedisPool.getResource();
+                Map<String, String> keyValue = jedis.hgetAll(this.options.hkey);
+                List<String> values = buildValuesFromMap(keys, keyValue);
+                return deserializeValues(keys, values);
+            } finally {
+                if (jedis != null) {
+                    jedisPool.returnResource(jedis);
+                }
+            }
+        }
+    }
+
+    private List<String> buildValuesFromMap(List<List<Object>> keys, Map<String, String> keyValue) {
+        List<String> values = new ArrayList<String>(keys.size());
+        for (List<Object> key : keys) {
+            String strKey = keyFactory.build(key);
+            String value = keyValue.get(strKey);
+            values.add(value);
+        }
+        return values;
+    }
+
+    private List<T> deserializeValues(List<List<Object>> keys, List<String> values) {
+        List<T> result = new ArrayList<T>(keys.size());
+        for (String value : values) {
+            if (value != null) {
+                result.add((T) serializer.deserialize(value.getBytes()));
+            } else {
+                result.add(null);
+            }
+        }
+        return result;
+    }
+
+    private String[] buildKeys(List<List<Object>> keys) {
+        String[] stringKeys = new String[keys.size()];
+        int index = 0;
+        for (List<Object> key : keys)
+            stringKeys[index++] = keyFactory.build(key);
+        return stringKeys;
+    }
+
+    public void multiPut(List<List<Object>> keys, List<T> vals) {
+        if (keys.size() == 0) {
+            return;
+        }
+
+        if (Strings.isNullOrEmpty(this.options.hkey)) {
+            Jedis jedis = null;
+            try {
+                jedis = jedisPool.getResource();
+                String[] keyValue = buildKeyValuesList(keys, vals);
+                jedis.mset(keyValue);
+            } finally {
+                if (jedis != null) {
+                    jedisPool.returnResource(jedis);
+                }
+            }
+        } else {
+            Jedis jedis = jedisPool.getResource();
+            try {
+                Pipeline pl = jedis.pipelined();
```

Comment:
@dashengju We can make buildKeyValuesMap() and use hmset instead of Pipeline & multiple hset.